dags/trending-snapshots-dag.py

```diff
@@ -0,0 +1,166 @@
+"""
+This script is used to execute a query on Snowflake and report the results to a 
+Synapse table. This DAG updates the Trending Snapshots Synapse table
+(https://www.synapse.org/Synapse:syn61597055/tables/) every month with the following metrics:
+
+- Top 10 Projects with the most unique users
+- Number of unique users
+- Last download date
+- Total data size (in GiB) for each project
+- Export date (the last time this table was updated)
+"""
+
+from airflow.decorators import dag
+
+from dataclasses import dataclass
+from datetime import date, datetime
+from typing import List
+
+import synapseclient
+from airflow.decorators import dag, task
+from airflow.models.param import Param
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+from orca.services.synapse import SynapseHook
+
+
+SYNAPSE_RESULTS_TABLE = "syn61932294"
+
+dag_params = {
+    "snowflake_conn_id": Param("SNOWFLAKE_SYSADMIN_PORTAL_RAW_CONN", type="string"),
+    "synapse_conn_id": Param("SYNAPSE_ORCA_SERVICE_ACCOUNT_CONN", type="string"),
+    "current_date": Param(date.today().strftime("%Y-%m-%d"), type="string"),
+    "time_window": Param("30d", type="string"),
+    }
+
+dag_config = {
+    # run on the 1st of every month at midnight
+    "schedule_interval": "0 0 1 * *",
+    "start_date": datetime(2024, 7, 1),
+    "catchup": False,
+    "default_args": {
+        "retries": 1,
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+
+@dataclass
+class SnapshotsMetric:
+    """
+    Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        total_data_size_in_pib: The size of the data hosted on Synapse in PiB
+        active_users_last_month: The number of active users last month
+        total_downloads_last_month: The total number of downloads by users last month
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    total_data_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_snapshots() -> None:
+    """
+    """
+
+    @task
+    def get_trending_snapshots(**context) -> List[SnapshotsMetric]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH RECENT_DOWNLOADS AS (
+                    SELECT *
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE 1=1
+                    AND RECORD_DATE > DATEADD(DAY, -28, '2024-06-26') 
```

Comment:
I'm changing this later. I'm trying to re-create the numbers in the proof-of-concept schema in the Google Sheets. Right now the numbers are _slightly_ different, if you run this on Snowsight, and my goal is to figure that out tomorrow.